File: backend/app/schemas/purchase.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class SupplierBase(BaseModel):
    supplier_name: str
    contact_person: str
    contact_phone: str
    address: str


class SupplierCreate(SupplierBase):
    pass


class SupplierUpdate(BaseModel):
    supplier_name: str | None = None
    contact_person: str | None = None
    contact_phone: str | None = None
    address: str | None = None


class SupplierOut(SupplierBase):
    model_config = ConfigDict(from_attributes=True)

    supplier_id: str


class IngredientBase(BaseModel):
    ingredient_name: str
    unit: str
    category: str
    shelf_life_days: int = Field(gt=0)


class IngredientCreate(IngredientBase):
    pass


class IngredientUpdate(BaseModel):
    ingredient_name: str | None = None
    unit: str | None = None
    category: str | None = None
    shelf_life_days: int | None = Field(default=None, gt=0)


class IngredientOut(IngredientBase):
    model_config = ConfigDict(from_attributes=True)

    ingredient_id: str


class PurchaseDetailBase(BaseModel):
    ingredient_id: str
    purchase_qty: Decimal = Field(gt=0)
    purchase_unit_price: Decimal = Field(gt=0)


class PurchaseDetailCreate(PurchaseDetailBase):
    pass


class PurchaseDetailOut(PurchaseDetailBase):
    model_config = ConfigDict(from_attributes=True)

    purchase_detail_id: str
    purchase_order_id: str


class PurchaseOrderBase(BaseModel):
    supplier_id: str
    order_date: date


class PurchaseOrderCreate(PurchaseOrderBase):
    order_status: str = "PENDING"
    details: list[PurchaseDetailCreate] = Field(min_length=1)


class PurchaseOrderUpdate(BaseModel):
    supplier_id: str | None = None
    order_date: date | None = None
    details: list[PurchaseDetailCreate] | None = None


class PurchaseOrderOut(PurchaseOrderBase):
    model_config = ConfigDict(from_attributes=True)

    purchase_order_id: str
    order_total_amount: Decimal
    order_status: str
    details: list[PurchaseDetailOut] = []

File: frontend/src/types/purchase.ts
export interface Supplier {
  supplier_id: string
  supplier_name: string
  contact_person: string
  contact_phone: string
  address: string
}

export interface SupplierCreate {
  supplier_name: string
  contact_person: string
  contact_phone: string
  address: string
}

export type SupplierUpdate = Partial<SupplierCreate>

export interface Ingredient {
  ingredient_id: string
  ingredient_name: string
  unit: string
  category: string
  shelf_life_days: number
}

export interface IngredientCreate {
  ingredient_name: string
  unit: string
  category: string
  shelf_life_days: number
}

export type IngredientUpdate = Partial<IngredientCreate>

export interface PurchaseDetail {
  purchase_detail_id?: string
  purchase_order_id?: string
  ingredient_id: string
  purchase_qty: number | string
  purchase_unit_price: number | string
}

export interface PurchaseOrder {
  purchase_order_id: string
  supplier_id: string
  order_date: string
  order_total_amount: number | string
  order_status: string
  details: PurchaseDetail[]
}

export interface PurchaseDetailCreate {
  ingredient_id: string
  purchase_qty: number
  purchase_unit_price: number
}

export interface PurchaseOrderCreate {
  supplier_id: string
  order_date: string
  order_status?: string
  details: PurchaseDetailCreate[]
}

export interface PurchaseOrderUpdate {
  supplier_id?: string
  order_date?: string
  details?: PurchaseDetailCreate[]
}
